timemachine/cpp/src/kernels/k_exchange.cu

```diff
@@ -0,0 +1,740 @@
+    const double *__restrict__ coords,
+    float *__restrict__ centroid);
+template void __global__ k_compute_centroid_of_atoms<double>(
+    const int num_atoms,
+    const int *__restrict__ atom_idxs,
+    const double *__restrict__ coords,
+    double *__restrict__ centroid);
+
+template <typename RealType>
+void __global__ k_split_mols_inner_outer(
+    const int num_molecules,
+    const int *__restrict__ atom_idxs,
+    const int *__restrict__ mol_offsets, // [num_molecules + 1]
+    const RealType *__restrict__ center, // [3] Center that determines inner vs outer
+    const RealType square_radius,        // squared radius from center that defines inner
+    const double *__restrict__ coords,   // [N, 3]
+    const double *__restrict__ box,      // [3, 3]
+    int *__restrict__ inner_count,       // [1]
+    int *__restrict__ inner_mols,        // [num_molecules]
+    int *__restrict__ outer_count,       // [1]
+    int *__restrict__ outer_mols         // [num_molecules]
+) {
+    int mol_idx = blockIdx.x * blockDim.x + threadIdx.x;
+
+    const RealType box_x = box[0 * 3 + 0];
+    const RealType box_y = box[1 * 3 + 1];
+    const RealType box_z = box[2 * 3 + 2];
+
+    const RealType inv_box_x = 1 / box_x;
+    const RealType inv_box_y = 1 / box_y;
+    const RealType inv_box_z = 1 / box_z;
+
+    const RealType target_x = center[0];
+    const RealType target_y = center[1];
+    const RealType target_z = center[2];
+
+    while (mol_idx < num_molecules) {
+        int mol_start = mol_offsets[mol_idx];
+        int mol_end = mol_offsets[mol_idx + 1];
+        int start_idx = atom_idxs[mol_start];
+        int num_atoms = mol_end - mol_start;
+
+        unsigned long long centroid_accum_x = 0;
+        unsigned long long centroid_accum_y = 0;
+        unsigned long long centroid_accum_z = 0;
+        for (int atom_idx = start_idx; atom_idx < start_idx + num_atoms; atom_idx++) {
+            centroid_accum_x += FLOAT_TO_FIXED<RealType>(coords[atom_idx * 3 + 0]);
+            centroid_accum_y += FLOAT_TO_FIXED<RealType>(coords[atom_idx * 3 + 1]);
+            centroid_accum_z += FLOAT_TO_FIXED<RealType>(coords[atom_idx * 3 + 2]);
+        }
+
+        RealType centroid_x = FIXED_TO_FLOAT<RealType>(centroid_accum_x) / static_cast<RealType>(num_atoms);
+        RealType centroid_y = FIXED_TO_FLOAT<RealType>(centroid_accum_y) / static_cast<RealType>(num_atoms);
+        RealType centroid_z = FIXED_TO_FLOAT<RealType>(centroid_accum_z) / static_cast<RealType>(num_atoms);
+
+        centroid_x -= target_x;
+        centroid_y -= target_y;
+        centroid_z -= target_z;
+
+        centroid_x -= box_x * nearbyint(centroid_x * inv_box_x);
+        centroid_y -= box_y * nearbyint(centroid_y * inv_box_y);
+        centroid_z -= box_z * nearbyint(centroid_z * inv_box_z);
+
+        RealType dist = (centroid_x * centroid_x) + (centroid_y * centroid_y) + (centroid_z * centroid_z);
+        // This is not deterministic in the ordering....
```

Comment:
Does this mean that the generated ordering of mols differ for the inner/outer mols? Does the energy computation differ if the ordering of the mols different?

Reply:
The ordering of the `inner_mols` and `outer_mols` is not deterministic due to the atomic adds to the counts (could do a cub IF computation like in the local MD code, but seemed more expensive). It doesn't change the energies but it does make the sampling non-deterministic so you can get different mols sampled.

Reply:
(resolved offline)

Reply:
Switched to partitioning in https://github.com/proteneer/timemachine/pull/1205/commits/4f2e9fc8f803866f44be56e208b400605bb52240 at @proteneer recommendation. Speeds things up by 40%